### ISSUE TITLE
dynamic imports with pending unresolved transitive dependencies can fail

### ISSUE THREAD

**Issue by xReveres** [denoland/deno#9929]
Hi.
I'm trying to speed up my framework, so I decided to rewrite the process of loading dynamic typescript files.
Old code:
```ts
let loadedFiles = [];
for await(const dirEntry of Deno.readDir( targetDirectory )){
    loadedFiles.push(await import(path.join(targetDirectory, dirEntry.name)));
}
```
This method takes a long time to load all files from directory. I need to wait 7-10sec to load all (precompiled) files.
Today I realized that I can load it like this:
```ts
let promises = [];
for await(const dirEntry of Deno.readDir( targetDirectory )){
    loadedFiles.push(await import(path.join(targetDirectory, dirEntry.name)));
}
await Promise.all(promises).then((loadedFiles)=>{ ... });
```
But deno throws error:
```sh
Cannot resolve module "./rsa_internal.ts" from "https://deno.land/x/god_crypto@v1.4.9/src/rsa/rsa_js.ts"
```
I tried to change lib but unfortunatelly also throws error:
```sh
Cannot resolve module "https://denopkg.com/chiefbiiko/std-encoding@v1.1.1/mod.ts" from "https://deno.land/x/hmac@v2.0.1/deps.ts"
```

**Comment by kitsonk** [comment]
You aren't adding anything to `promises` in the second example.  What do you expect to happen?

**Comment by kturcios** [comment]
What @kitsonk said, but you also shouldn't need to use 'await' for your imports inside the for loop because Promise.all expects an array of promises

**Comment by xReveres** [comment]
@kitsonk @kturcios sorry for late answer, yes my mistake. This is just example of what I'm trying to do.
Better example:
```ts
const promises = [];
for await(const dirEntry of Deno.readDir(filesDirectory)){
    if(!dirEntry.isFile || dirEntry.name.toLowerCase() == 'mod.ts' || dirEntry.name.toLowerCase() == 'deps.ts')
        continue;
    const name = dirEntry.name.substr(0, dirEntry.name.length - 3);
    const filepath = path.join(filesDirectory, dirEntry.name);
    promises.push(this.loadController(name, filepath));
}
await Promise.all(promises).then((loadedFiles) => { ... });

async loadFile(name: string, filePath: string){
    const file = await import(filepath);
    return new file[name]();
}
```

**Comment by xReveres** [comment]
@kitsonk Okay I have example to reproduce:
Directories structure:
![image](https://user-images.githubusercontent.com/8580908/113002197-d6711800-9171-11eb-9995-9db9eea2aead.png)
**bug.ts**
```ts
const DIRECTORY = './files';

async function loadFile(name: string, filepath: string){
    const file = await import(filepath);
    return new file[name]();
}

const promises: Promise<any>[] = [];
console.log('Reading directory');
for await(const dirEntry of Deno.readDir(DIRECTORY)){
    const name = dirEntry.name.substr(0, dirEntry.name.length - 3);
    promises.push(loadFile(name, DIRECTORY+'/'+dirEntry.name));
}

console.log('Loading files');
await Promise.all(promises).then((files)=>{
    console.log('loaded files:', files);
});
console.log('End');

export {} //For working for await
```
**deps.ts**
```ts
export { hmac } from "https://deno.land/x/hmac@v2.0.1/mod.ts";
```
**files/LoadFirst.ts**
```ts
import { hmac } from "https://deno.land/x/hmac@v2.0.1/mod.ts";

export class LoadFirst{
    constructor(){
        console.log('--- First');
        hmac('sha1', 'asd', '123');
    }
}
```
**files/LoadSecond.ts**
```ts
export class LoadSecond{
    constructor(){
        console.log('--- Second');
    }
}
```
**files/LoadThird.ts**
```ts
export class LoadThird{
    constructor(){
        console.log('--- Third');
    }
}
```
**files/LoadFourth.ts**
```ts
import { hmac } from '../deps.ts'

export class LoadFourth{
    constructor(){
        console.log('--- Fourth');
        hmac('sha1', 'mykey', 'my message')
    }
}
```
Run CMD:
```sh
deno run --allow-read bug.ts
```
And boom, throws error:
```ts
Reading directory
Loading files
--- Second
--- Third
--- First
error: Uncaught (in promise) TypeError: Cannot resolve module "https://denopkg.com/chiefbiiko/sha1@v2.0.0/mod.ts" from "https://deno.land/x/hmac@v2.0.1/deps.ts"
    const file = await import(filepath);
```
```sh
deno --version
deno 1.8.2 (release, x86_64-unknown-linux-gnu)
v8 9.0.257.3
typescript 4.2.2
```

**Comment by kitsonk** [comment]
@bartlomieju I think we might have a real issue here... I am not sure if it is in the file fetcher or the module loader yet, and I suspect there is a far simpler reproduction, but I think when there are two modules being dynamically imported and the modules and their dependencies haven't been fully resolved, and the depend on the same transitive dependencies, we can fail...

**Comment by lucacasonato** [comment]
I have actually run into this too previously. To fix it I had to dynamic import one after the other: https://github.com/lucacasonato/dext.ts/commit/f82e411969f20563e3dcf338dce567aa213b0672. It only started happening pretty recently (possibly starting 1.8.0?).

**Comment by bartlomieju** [comment]
Probably related to https://github.com/denoland/deno/issues/3736 ?

**Comment by kitsonk** [comment]
> This can happen if there are circular dependencies in dynamic imports

It isn't a circular dependency, there are two dynamic imports that have been imported, but their promises haven't been resolved in the isolate yet, so a duplicated dependency if you will.

**Comment by bartlomieju** [comment]
Okay, thanks for the info, I need to dig deeper to pinpoint the problem.

**Comment by nayeemrmn** [comment]
See #9982.

**Comment by nayeemrmn** [comment]
Okay #9982 wasn't quite related, I suppose this is a bug in the file fetcher. Smaller repro:

`temp.js`:
```js
import("./temp2.js");
import("./temp3.js");
```
`temp2.js`:
```js
import "https://deno.land/x/hmac@v2.0.1/mod.ts";
```
`temp3.js`:
```js
import "./temp4.js";
```
`temp4.js`:
```js
import "https://deno.land/x/hmac@v2.0.1/mod.ts";
```
`deno run --allow-read temp.js`:
```
error: Uncaught (in promise) TypeError: Cannot resolve module "https://denopkg.com/chiefbiiko/sha1@v2.0.0/mod.ts" from "https://deno.land/x/hmac@v2.0.1/deps.ts"
```

**Comment by bartlomieju** [comment]
@nayeemrmn then it's the same issue as #3736. It's actual bug in module loading implementation in `deno_core`.

**Comment by bartlomieju** [comment]
Fixed by #11089
